File: PokerWorld/src/poker/PokerGame.java
package poker;

import java.util.List;

import poker.observer.PokerGameObserver;

public class PokerGame implements IPokerGame
{
    private final PokerGameObserver m_gameObserver;
    private final PokerTableInfo m_pokerTable;
    private TypePokerGameState m_currentGameState;
    private TypePokerGameRoundState m_currentGameRoundState;
    private final int m_WaitingTimeAfterPlayerAction;
    private final int m_WaitingTimeAfterBoardDealed;
    private final int m_WaitingTimeAfterPotWon;
    
    private final AbstractPokerDealer m_pokerDealer;
    
    public PokerGame()
    {
        this(new RandomPokerDealer());
    }
    
    public PokerGame(PokerTableInfo info, int wtaPlayerAction, int wtaBoardDealed, int wtaPotWon)
    {
        this(new RandomPokerDealer(), info, wtaPlayerAction, wtaBoardDealed, wtaPotWon);
    }
    
    public PokerGame(AbstractPokerDealer dealer)
    {
        this(new RandomPokerDealer(), new PokerTableInfo(), 0, 0, 0);
    }
    
    public PokerGame(AbstractPokerDealer dealer, PokerTableInfo info, int wtaPlayerAction, int wtaBoardDealed, int wtaPotWon)
    {
        m_pokerDealer = dealer;
        m_gameObserver = new PokerGameObserver();
        m_pokerTable = info;
        m_currentGameState = TypePokerGameState.INIT;
        m_WaitingTimeAfterPlayerAction = wtaPlayerAction;
        m_WaitingTimeAfterBoardDealed = wtaBoardDealed;
        m_WaitingTimeAfterPotWon = wtaPotWon;
    }
    
    public void start()
    {
        setCurrentGameState(TypePokerGameState.PLAYERS_WAITING);
    }
    
    public TypePokerGameState getCurrentGameState()
    {
        return m_currentGameState;
    }
    
    public TypePokerGameRound getCurrentGameRound()
    {
        return m_pokerTable.getCurrentGameRound();
    }
    
    public TypePokerGameRoundState getCurrentGameRoundState()
    {
        return m_currentGameRoundState;
    }
    
    private void setCurrentGameState(TypePokerGameState newGS)
    {
        
        final TypePokerGameState oldGS = m_currentGameState;
        
        if (m_currentGameState == TypePokerGameState.END)
        {
            return;
        }
        
        if (newGS.ordinal() - oldGS.ordinal() != 1)
        {
            return;
        }
        
        m_currentGameState = newGS;
        
        switch (m_currentGameState)
        {
            case INIT:
                break;
            case PLAYERS_WAITING:
                TryToBegin();
                break;
            case BLIND_WAITING:
                m_pokerTable.setCurrentHigherBet(0);
                break;
            case PLAYING:
                m_pokerTable.setCurrentGameRound(TypePokerGameRound.PREFLOP);
                m_currentGameRoundState = TypePokerGameRoundState.CARDS;
                startRound();
                break;
            case SHOWDOWN:
                showAllCards();
                break;
            case DECIDE_WINNERS:
                decideWinners();
                break;
            case MONEY_DISTRIBUTION:
                distributeMoney();
                break;
            case END:
                m_gameObserver.everythingEnded();
                break;
            
        }
    }
    
    private void setCurrentGameRound(TypePokerGameRound newGR)
    {
        
        final TypePokerGameRound oldGR = m_pokerTable.getCurrentGameRound();
        
        if (m_currentGameState != TypePokerGameState.PLAYING)
        {
            return;
        }
        
        if (newGR.ordinal() - oldGR.ordinal() != 1)
        {
            return;
        }
        
        m_currentGameRoundState = TypePokerGameRoundState.CARDS;
        m_pokerTable.setCurrentPlayerNoSeat(m_pokerTable.getCurrentDealerNoSeat());
        m_pokerTable.setCurrentGameRound(newGR);
        startRound();
    }
    
    private void setCurrentGameRoundState(TypePokerGameRoundState newGRS)
    {
        
        final TypePokerGameRoundState oldGRS = m_currentGameRoundState;
        
        if (m_currentGameState != TypePokerGameState.PLAYING)
        {
            return;
        }
        
        if (newGRS.ordinal() - oldGRS.ordinal() != 1)
        {
            return;
        }
        
        m_currentGameRoundState = newGRS;
        startRound();
    }
    
    public PokerGameObserver getGameObserver()
    {
        return m_gameObserver;
    }
    
    public PokerTableInfo getPokerTable()
    {
        return m_pokerTable;
    }
    
    public boolean isRunning()
    {
        return m_currentGameState != TypePokerGameState.END;
    }
    
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // ///////////////// COMMUNICATION ///////////////////////////
    // /////////////////////////////////
    public boolean joinGame(PokerPlayerInfo p)
    {
        if (m_currentGameState == TypePokerGameState.INIT || m_currentGameState == TypePokerGameState.END)
        {
            System.err.println("Bad timing:" + m_currentGameState);
            return false;
        }
        
        return m_pokerTable.joinTable(p);
    }
    
    public void sitInGame(PokerPlayerInfo p)
    {
        m_gameObserver.playerJoined(p);
        if (m_currentGameState == TypePokerGameState.PLAYERS_WAITING)
        {
            TryToBegin();
        }
    }
    
    public boolean leaveGame(PokerPlayerInfo p)
    {
        if (m_pokerTable.leaveTable(p))
        {
            m_gameObserver.playerLeaved(p);
            return true;
        }
        return false;
    }
    
    public boolean playMoney(PokerPlayerInfo p, int amount)
    {
        final int amnt = Math.min(amount, p.getCurrentSafeMoneyAmount());
        System.out.println(p.getPlayerName() + " is playing " + amnt + " money on state: " + m_currentGameState);
        if (m_currentGameState == TypePokerGameState.BLIND_WAITING)
        {
            System.out.println(p.getPlayerName() + " is putting blind of " + amnt);
            System.out.println("Total still needed is " + m_pokerTable.getTotalBlindNeeded());
            final int needed = m_pokerTable.blindNeeded(p);
            if (amnt != needed)
            {
                if (p.canBet(amnt + 1))
                {
                    System.err.println(p.getPlayerName() + " needed to put " + needed);
                    return false;
                }
                else
                {
                    System.out.println("So ... All-In !");
                    p.setAllIn();
                    m_pokerTable.incNbAllIn();
                    m_pokerTable.addAllInCap(p.getCurrentBetMoneyAmount() + amnt);
                }
            }
            if (!p.tryBet(amnt))
            {
                System.err.println(p.getPlayerName() + " just .. can't !! ");
                return false;
            }
            m_pokerTable.incTotalPotAmount(amnt);
            m_gameObserver.playerMoneyChanged(p);
            m_pokerTable.setBlindNeeded(p, 0);
            if (amnt == m_pokerTable.getSmallBlindAmount())
            {
                m_gameObserver.playerActionTaken(p, TypePokerGameAction.SMALL_BLIND_POSTED, amnt);
            }
            else
            {
                m_gameObserver.playerActionTaken(p, TypePokerGameAction.BIG_BLIND_POSTED, amnt);
            }
            m_pokerTable.setTotalBlindNeeded(m_pokerTable.getTotalBlindNeeded() - needed);
            if (m_pokerTable.getTotalBlindNeeded() == 0)
            {
                setCurrentGameState(TypePokerGameState.PLAYING);
            }
            if (amnt > m_pokerTable.getCurrentHigherBet())
            {
                m_pokerTable.setCurrentHigherBet(amnt);
            }
            return true;
        }
        
        else if (m_currentGameState == TypePokerGameState.PLAYING && m_currentGameRoundState == TypePokerGameRoundState.BETTING)
        {
            System.out.println("Currently, we need " + m_pokerTable.getCallAmount(p) + " minimum money from this player");
            if (p.getCurrentTablePosition() != m_pokerTable.getCurrentPlayerNoSeat())
            {
                System.err.println("BUT SCREW YOU, IT'S NOT YOUR TURN !!!!!");
                return false;
            }
            
            if (amnt == -1)
            {
                System.out.println("So ... the little girl folds !");
                foldPlayer(p);
                continueBettingRound();
                return true;
            }
            int amntNeeded = m_pokerTable.getCallAmount(p);
            if (amnt < amntNeeded)
            {
                if (p.canBet(amnt + 1))
                {
                    System.err.println("BUT SCREW YOU, IT'S NOT ENOUGH !!!!!");
                    return false;
                }
                System.out.println("So ... All-In ! getCurrentBetMoneyAmount: " + p.getCurrentBetMoneyAmount());
                amntNeeded = amnt;
                p.setAllIn();
                m_pokerTable.incNbAllIn();
                m_pokerTable.addAllInCap(p.getCurrentBetMoneyAmount() + amnt);
            }
            if (!p.tryBet(amnt))
            {
                System.err.println("BUT SCREW YOU, YOU JUST CAN'T !!!!!");
                return false;
            }
            m_gameObserver.playerMoneyChanged(p);
            if (amnt == amntNeeded)
            {
                System.out.println("Will call with $" + amnt);
                m_pokerTable.incTotalPotAmount(amnt);
                callPlayer(p, amnt);
                continueBettingRound();
                return true;
            }
            System.out.println("Will raise with $" + amnt);
            m_pokerTable.incTotalPotAmount(amnt);
            raisePlayer(p, amnt);
            playNext();
            return true;
        }
        System.err.println("BUT WE DON'T CARE !!!!!");
        return false;
    }
    
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // /////////////////////////////////
    // ////// PRIV /////////////////////
    // /////////////////////////////////
    
    private void startRound()
    {
        switch (m_currentGameRoundState)
        {
            case CARDS:
                startCardRound();
                break;
            case BETTING:
                startBettingRound();
                break;
            case CUMUL:
                startCumulRound();
                break;
        }
    }
    
    private void showAllCards()
    {
        for (final PokerPlayerInfo p : m_pokerTable.getPlayers())
        {
            if (p.isPlaying() || p.isAllIn())
            {
                p.setShowingCards(true);
                m_gameObserver.playerHoleCardsChanged(p);
            }
        }
        setCurrentGameState(TypePokerGameState.DECIDE_WINNERS);
    }
    
    private void startCumulRound()
    {
        m_pokerTable.managePotsRoundEnd();
        m_gameObserver.gameBettingRoundEnded(m_pokerTable.getCurrentGameRound());
        if (m_pokerTable.getNbPlaying() == 1 && m_pokerTable.getNbAllIn() == 0)
        {
            setCurrentGameState(TypePokerGameState.SHOWDOWN);
        }
        else
        {
            switch (m_pokerTable.getCurrentGameRound())
            {
                case PREFLOP:
                    setCurrentGameRound(TypePokerGameRound.FLOP);
                    break;
                case FLOP:
                    setCurrentGameRound(TypePokerGameRound.TURN);
                    break;
                case TURN:
                    setCurrentGameRound(TypePokerGameRound.RIVER);
                    break;
                case RIVER:
                    setCurrentGameState(TypePokerGameState.SHOWDOWN);
                    break;
            }
        }
    }
    
    private void startBettingRound()
    {
        m_gameObserver.gameBettingRoundStarted();
        m_pokerTable.setNbPlayed(0);
        waitALittle(m_WaitingTimeAfterBoardDealed);
        continueBettingRound();
    }
    
    private void startCardRound()
    {
        switch (m_pokerTable.getCurrentGameRound())
        {
            case PREFLOP:
                m_pokerTable.setCurrentPlayerNoSeat(m_pokerTable.getCurrentBigBlindNoSeat());
                dealHole();
                break;
            case FLOP:
                dealFlop();
                break;
            case TURN:
                dealTurn();
                break;
            case RIVER:
                dealRiver();
                break;
        }
        setCurrentGameRoundState(TypePokerGameRoundState.BETTING);
    }
    
    private void dealRiver()
    {
        m_pokerTable.addBoardCard(m_pokerDealer.dealRiver());
    }
    
    private void dealTurn()
    {
        m_pokerTable.addBoardCard(m_pokerDealer.dealTurn());
    }
    
    private void dealFlop()
    {
        m_pokerTable.addBoardCards(m_pokerDealer.dealFlop());
    }
    
    private void dealHole()
    {
        for (final PokerPlayerInfo p : m_pokerTable.getPlayingPlayers())
        {
            p.setHand(m_pokerDealer.dealHoles(p));
            m_gameObserver.playerHoleCardsChanged(p);
        }
    }
    
    private void foldPlayer(PokerPlayerInfo p)
    {
        p.setFolded();
        m_gameObserver.playerActionTaken(p, TypePokerGameAction.FOLDED, -1);
    }
    
    private void TryToBegin()
    {
        System.out.print("Trying to begin ...");
        m_pokerTable.getAndSetNbPlayingPlayers();
        if (m_pokerTable.getNbPlaying() > 1)
        {
            System.out.println(" yep ! do it !");
            m_pokerTable.initTable();
            m_pokerDealer.freshDeck();
            setCurrentGameState(TypePokerGameState.BLIND_WAITING);
            m_gameObserver.gameBlindsNeeded();
        }
        else
        {
            System.out.println(" just too bad :(");
            m_pokerTable.setCurrentDealerNoSeat(-1);
            m_pokerTable.setCurrentSmallBlindNoSeat(-1);
            m_pokerTable.setCurrentSmallBlindNoSeat(-1);
        }
    }
    
    private void callPlayer(PokerPlayerInfo p, int played)
    {
        m_pokerTable.incNbPlayed();
        m_gameObserver.playerActionTaken(p, TypePokerGameAction.CALLED, played);
    }
    
    private void raisePlayer(PokerPlayerInfo p, int played)
    {
        m_pokerTable.setNbPlayed(1);
        m_pokerTable.setCurrentHigherBet(p.getCurrentBetMoneyAmount());
        m_gameObserver.playerActionTaken(p, TypePokerGameAction.RAISED, played);
    }
    
    private void continueBettingRound()
    {
        // TODO: RICK: Semble que 2/2 all-in du meme montant au preflop chie: continue a jouer
        waitALittle(m_WaitingTimeAfterPlayerAction);
        if (m_pokerTable.getNbPlaying() == 1 || m_pokerTable.getNbPlayed() >= m_pokerTable.getNbPlaying())
        {
            endBettingRound();
        }
        else
        {
            playNext();
        }
    }
    
    private void waitALittle(int waitingTime)
    {
        try
        {
            Thread.sleep(waitingTime);
        }
        catch (final InterruptedException e)
        {
            e.printStackTrace();
        }
    }
    
    private void endBettingRound()
    {
        setCurrentGameRoundState(TypePokerGameRoundState.CUMUL);
    }
    
    private void playNext()
    {
        final PokerPlayerInfo player = m_pokerTable.nextPlayingPlayer(m_pokerTable.getCurrentPlayerNoSeat());
        m_pokerTable.setCurrentPlayerNoSeat(player.getCurrentTablePosition());
        m_gameObserver.playerActionNeeded(player);
    }
    
    private void distributeMoney()
    {
        for (final PokerMoneyPot pot : m_pokerTable.getPots())
        {
            final List<PokerPlayerInfo> players = pot.getAttachedPlayers();
            if (players.size() > 0)
            {
                final int wonAmount = pot.getAmount() / players.size();
                for (final PokerPlayerInfo p : players)
                {
                    p.incCurrentSafeMoneyAmount(wonAmount);
                    m_gameObserver.playerMoneyChanged(p);
                    m_gameObserver.playerWonPot(p, pot, wonAmount);
                    
                    waitALittle(m_WaitingTimeAfterPotWon);
                }
            }
            else
            {
                System.err.println(">> POT SANS PLAYER: ANORMAL !!");
            }
        }
        m_gameObserver.gameEnded();
        m_currentGameState = TypePokerGameState.PLAYERS_WAITING;
        TryToBegin();
    }
    
    private void decideWinners()
    {
        m_pokerTable.cleanPotsForWinning();
        setCurrentGameState(TypePokerGameState.MONEY_DISTRIBUTION);
    }
}

File: C#Integration/PokerWorld/Game/PokerGame.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

namespace PokerWorld.Game
{
    public class PokerGame : IPokerGame
    {
        // Global States of the Game
        public enum TypeState
        {
            Init,
            WaitForPlayers,
            WaitForBlinds,
            Playing,
            Showdown,
            DecideWinners,
            DistributeMoney,
            End
        }

        // States of the Game in each Round
        public enum TypeRoundState
        {
            Cards,
            Betting,
            Cumul
        }

        public event EventHandler EverythingEnded = delegate { };
        public event EventHandler GameBlindNeeded = delegate { };
        public event EventHandler GameEnded = delegate { };
        public event EventHandler GameGenerallyUpdated = delegate { };
        public event EventHandler<RoundEventArgs> GameBettingRoundStarted = delegate { };
        public event EventHandler<RoundEventArgs> GameBettingRoundEnded = delegate { };
        public event EventHandler<PlayerInfoEventArgs> PlayerJoined = delegate { };
        public event EventHandler<PlayerInfoEventArgs> PlayerLeaved = delegate { };
        public event EventHandler<HistoricPlayerInfoEventArgs> PlayerActionNeeded = delegate { };
        public event EventHandler<PlayerInfoEventArgs> PlayerMoneyChanged = delegate { };
        public event EventHandler<PlayerInfoEventArgs> PlayerHoleCardsChanged = delegate { };
        public event EventHandler<PlayerActionEventArgs> PlayerActionTaken = delegate { };
        public event EventHandler<PotWonEventArgs> PlayerWonPot = delegate { };

        private readonly TableInfo m_Table; // La table
        private readonly AbstractDealer m_Dealer; // Dealer

        // WAITING TIME
        private readonly int m_WaitingTimeAfterPlayerAction; // Attente apres chaque player action (ms)
        private readonly int m_WaitingTimeAfterBoardDealed; // Attente apres chaque board dealed (ms)
        private readonly int m_WaitingTimeAfterPotWon; // Attente apres chaque pot won ! (ms)

        // STATES
        private TypeState m_State; // L'etat global de la game
        private TypeRoundState m_RoundState; // L'etat de la game pour chaque round

        public TableInfo Table
        {
            get { return m_Table; }
        }
        public TypeRound Round
        {
            get { return m_Table.Round; }
        }

        public bool IsRunning
        {
            get { return m_State != TypeState.End; }
        }

        public string Encode
        {
            get
            {
                // 0 : Assume que les game sont en real money (1)
                // 1 : Assume que c'est tlt du Texas Hold'em (0)
                // 2 : Assume que c'Est tlt des Ring game (0)
                // 3 : Assume que c'Est tlt du NoLimit (0)
                // 4 : GameRound (0,1,2,3)
                return string.Format("{0}{1}{2}{3}{4}",1,0,0,0,(int)m_Table.Round);
            }
        }

        public PokerGame()
            : this(new RandomDealer())
        {
        }
        public PokerGame(AbstractDealer dealer) :
            this(new RandomDealer(), new TableInfo(), 0, 0, 0)
        {
        }

        public PokerGame(TableInfo table, int wtaPlayerAction, int wtaBoardDealed, int wtaPotWon)
            : this(new RandomDealer(), table, wtaPlayerAction, wtaBoardDealed, wtaPotWon)
        {
        }

        public PokerGame(AbstractDealer dealer, TableInfo table, int wtaPlayerAction, int wtaBoardDealed, int wtaPotWon)
        {
            m_Dealer = dealer;
            m_Table = table;
            m_State = TypeState.Init;
            m_RoundState = TypeRoundState.Cards;
            m_WaitingTimeAfterPlayerAction = wtaPlayerAction;
            m_WaitingTimeAfterBoardDealed = wtaBoardDealed;
            m_WaitingTimeAfterPotWon = wtaPotWon;
        }

        private void NextState(TypeState state)
        {
            TypeState oldState = m_State;

            if (m_State == TypeState.End)
                return;

            if ((int)state - (int)oldState != 1)
                return;

            m_State = state;

            switch (m_State)
            {
                case TypeState.Init:
                    break;
                case TypeState.WaitForPlayers:
                    TryToBegin();
                    break;
                case TypeState.WaitForBlinds:
                    m_Table.HigherBet = 0;
                    break;
                case TypeState.Playing:
                    m_Table.Round = TypeRound.Preflop;
                    m_RoundState = TypeRoundState.Cards;
                    StartRound();
                    break;
                case TypeState.Showdown:
                    ShowAllCards();
                    break;
                case TypeState.DecideWinners:
                    DecideWinners();
                    break;
                case TypeState.DistributeMoney:
                    DistributeMoney();
                    break;
                case TypeState.End:
                    EverythingEnded(this, new EventArgs());
                    break;
            }
        }
        private void NextRound(TypeRound round)
        {

            TypeRound oldRound = m_Table.Round;

            if (m_State != TypeState.Playing)
                return;

            if ((int)round - (int)oldRound != 1)
                return;

            m_RoundState = TypeRoundState.Cards;
            m_Table.NoSeatCurrPlayer = m_Table.NoSeatDealer;
            m_Table.Round = round;
            StartRound();
        }
        private void NextRoundState(TypeRoundState roundState)
        {

            TypeRoundState oldRoundState = m_RoundState;

            if (m_State != TypeState.Playing)
                return;

            if ((int)roundState - (int)oldRoundState != 1)
                return;

            m_RoundState = roundState;
            StartRound();
        }

        public void Start()
        {
            NextState(TypeState.WaitForPlayers);
        }

        public bool JoinGame(PlayerInfo p)
        {
            if (m_State == TypeState.Init || m_State == TypeState.End)
            {
                Console.WriteLine("Bad timing:" + m_State);
                return false;
            }

            return m_Table.JoinTable(p);
        }

        public void SitInGame(PlayerInfo p)
        {
            PlayerJoined(this, new PlayerInfoEventArgs(p));
            if (m_State == TypeState.WaitForPlayers)
                TryToBegin();
        }

        public bool LeaveGame(PlayerInfo p)
        {
            if (m_Table.LeaveTable(p))
            {
                PlayerLeaved(this, new PlayerInfoEventArgs(p));
                if (m_Table.NbUsedSeats == 0)
                    m_State = TypeState.End;
                return true;
            }
            return false;
        }

        public bool PlayMoney(PlayerInfo p, int amount)
        {
            int amnt = Math.Min(amount, p.MoneySafeAmnt);
            Console.WriteLine(p.Name + " is playing " + amnt + " money on state: " + m_State);
            if (m_State == TypeState.WaitForBlinds)
            {
                Console.WriteLine("Total blinds needed is " + m_Table.TotalBlindNeeded);
                Console.WriteLine(p.Name + " is putting blind of " + amnt);
                int needed = m_Table.GetBlindNeeded(p);
                if (amnt != needed)
                {
                    if (p.CanBet(amnt + 1))
                    {
                        Console.WriteLine(p.Name + " needed to put " + needed);
                        return false;
                    }
                    else
                    {
                        Console.WriteLine("So ... All-In !");
                        p.IsAllIn = true;
                        m_Table.NbAllIn++;
                        m_Table.AddAllInCap(p.MoneyBetAmnt + amnt);
                    }
                }
                if (!p.TryBet(amnt))
                {
                    Console.WriteLine(p.Name + " just .. can't !! ");
                    return false;
                }
                m_Table.TotalPotAmnt += amnt;
                PlayerMoneyChanged(this, new PlayerInfoEventArgs(p));
                m_Table.AddBlindNeeded(p, 0);

                if (amnt == m_Table.SmallBlindAmnt)
                    PlayerActionTaken(this, new PlayerActionEventArgs(p, TypeAction.PostSmallBlind, amnt));
                else
                    PlayerActionTaken(this, new PlayerActionEventArgs(p, TypeAction.PostBigBlind, amnt));

                m_Table.TotalBlindNeeded -= needed;

                if (m_Table.TotalBlindNeeded == 0)
                    NextState(TypeState.Playing);

                if (amnt > m_Table.HigherBet)
                    m_Table.HigherBet = amnt;

                Console.WriteLine("Total blinds still needed is " + m_Table.TotalBlindNeeded);
                return true;
            }

            else if (m_State == TypeState.Playing && m_RoundState == TypeRoundState.Betting)
            {
                Console.WriteLine("Currently, we need " + m_Table.CallAmnt(p) + " minimum money from this player");
                if (p.NoSeat != m_Table.NoSeatCurrPlayer)
                {
                    Console.WriteLine("BUT SCREW YOU, IT'S NOT YOUR TURN !!!!!");
                    return false;
                }

                if (amnt == -1)
                {
                    Console.WriteLine("So ... the little girl folds !");
                    FoldPlayer(p);
                    ContinueBettingRound();
                    return true;
                }
                int amntNeeded = m_Table.CallAmnt(p);
                if (amnt < amntNeeded)
                {
                    if (p.CanBet(amnt + 1))
                    {
                        Console.WriteLine("BUT SCREW YOU, IT'S NOT ENOUGH !!!!!");
                        return false;
                    }
                    amntNeeded = amnt;
                }
                if (!p.TryBet(amnt))
                {
                    Console.WriteLine("BUT SCREW YOU, YOU JUST CAN'T !!!!!");
                    return false;
                }
                PlayerMoneyChanged(this, new PlayerInfoEventArgs(p));
                if (p.MoneySafeAmnt == 0)
                {
                    Console.WriteLine("So ... All-In ! getCurrentBetMoneyAmount: " + p.MoneyBetAmnt);
                    p.IsAllIn = true;
                    m_Table.NbAllIn++;
                    m_Table.AddAllInCap(p.MoneyBetAmnt);
                }
                if (amnt == amntNeeded)
                {
                    Console.WriteLine("Will call with $" + amnt);
                    m_Table.TotalPotAmnt += amnt;
                    CallPlayer(p, amnt);
                    ContinueBettingRound();
                    return true;
                }
                Console.WriteLine("Will raise with $" + amnt);
                m_Table.TotalPotAmnt += amnt;
                RaisePlayer(p, amnt);
                ContinueBettingRound();
                return true;
            }
            Console.WriteLine("BUT WE DON'T CARE !!!!!");
            return false;
        }
        private void StartRound()
        {
            switch (m_RoundState)
            {
                case TypeRoundState.Cards:
                    StartCardRound();
                    break;
                case TypeRoundState.Betting:
                    StartBettingRound();
                    break;
                case TypeRoundState.Cumul:
                    StartCumulRound();
                    break;
            }
        }
        private void StartCumulRound()
        {
            m_Table.ManagePotsRoundEnd();
            GameBettingRoundEnded(this, new RoundEventArgs(m_Table.Round));
            if (m_Table.NbPlayingAndAllIn <= 1)
            {
                NextState(TypeState.Showdown);
            }
            else
            {
                switch (m_Table.Round)
                {
                    case TypeRound.Preflop:
                        NextRound(TypeRound.Flop);
                        break;
                    case TypeRound.Flop:
                        NextRound(TypeRound.Turn);
                        break;
                    case TypeRound.Turn:
                        NextRound(TypeRound.River);
                        break;
                    case TypeRound.River:
                        NextState(TypeState.Showdown);
                        break;
                }
            }
        }
        private void StartBettingRound()
        {
            GameBettingRoundStarted(this, new RoundEventArgs(m_Table.Round));
            m_Table.NbPlayed = 0;
            m_Table.NoSeatLastRaise = m_Table.GetPlayingPlayerNextTo(m_Table.NoSeatCurrPlayer).NoSeat;
            WaitALittle(m_WaitingTimeAfterBoardDealed);

            if (m_Table.NbPlaying <= 1)
                EndBettingRound();
            else
                ContinueBettingRound();
        }
        private void StartCardRound()
        {
            switch (m_Table.Round)
            {
                case TypeRound.Preflop:
                    m_Table.NoSeatCurrPlayer = m_Table.NoSeatBigBlind;
                    DealHole();
                    break;
                case TypeRound.Flop:
                    DealFlop();
                    break;
                case TypeRound.Turn:
                    DealTurn();
                    break;
                case TypeRound.River:
                    DealRiver();
                    break;
            }
            NextRoundState(TypeRoundState.Betting);
        }
        private void DealRiver()
        {
            m_Table.AddCards(m_Dealer.DealRiver());
        }
        private void DealTurn()
        {
            m_Table.AddCards(m_Dealer.DealTurn());
        }
        private void DealFlop()
        {
            m_Table.AddCards(m_Dealer.DealFlop());
        }
        private void DealHole()
        {
            foreach (PlayerInfo p in m_Table.PlayingPlayers)
            {
                p.Cards = m_Dealer.DealHoles(p);
                PlayerHoleCardsChanged(this, new PlayerInfoEventArgs(p));
            }
        }
        private void ShowAllCards()
        {
            foreach (PlayerInfo p in m_Table.Players)
                if (p.IsPlaying || p.IsAllIn)
                {
                    p.IsShowingCards = true;
                    PlayerHoleCardsChanged(this, new PlayerInfoEventArgs(p));
                }
            NextState(TypeState.DecideWinners);
        }
        private void FoldPlayer(PlayerInfo p)
        {
            p.IsPlaying = false;
            WaitALittle(m_WaitingTimeAfterPlayerAction);
            PlayerActionTaken(this, new PlayerActionEventArgs(p, TypeAction.Fold, -1));
        }
        private void CallPlayer(PlayerInfo p, int played)
        {
            m_Table.NbPlayed++;
            WaitALittle(m_WaitingTimeAfterPlayerAction);
            PlayerActionTaken(this, new PlayerActionEventArgs(p, TypeAction.Call, played));
        }
        private void RaisePlayer(PlayerInfo p, int played)
        {
            int count = m_Table.NbAllIn;
            if (!p.IsAllIn)
                count++;
            m_Table.NoSeatLastRaise = p.NoSeat;
            m_Table.NbPlayed = count;
            m_Table.HigherBet = p.MoneyBetAmnt;
            WaitALittle(m_WaitingTimeAfterPlayerAction);
            PlayerActionTaken(this, new PlayerActionEventArgs(p, TypeAction.Raise, played));
        }
        private void ContinueBettingRound()
        {
            if (m_Table.NbPlayingAndAllIn == 1 || m_Table.NbPlayed >= m_Table.NbPlayingAndAllIn)
                EndBettingRound();
            else
                PlayNext();
        }
        private void EndBettingRound()
        {
            NextRoundState(TypeRoundState.Cumul);
        }
        private void PlayNext()
        {
            PlayerInfo old = m_Table.GetPlayer(m_Table.NoSeatCurrPlayer);
            PlayerInfo player = m_Table.GetPlayingPlayerNextTo(m_Table.NoSeatCurrPlayer);
            m_Table.NoSeatCurrPlayer = player.NoSeat;
            PlayerActionNeeded(this, new HistoricPlayerInfoEventArgs(player,old));
        }
        private void DistributeMoney()
        {
            foreach (MoneyPot pot in m_Table.Pots)
            {
                PlayerInfo[] players = pot.AttachedPlayers;
                if (players.Length > 0)
                {
                    int wonAmount = pot.Amount / players.Length;
                    if (wonAmount > 0)
                    {
                        foreach (PlayerInfo p in players)
                        {
                            p.MoneySafeAmnt += wonAmount;
                            PlayerMoneyChanged(this, new PlayerInfoEventArgs(p));
                            PlayerWonPot(this, new PotWonEventArgs(p, pot.Id, wonAmount));
                            WaitALittle(m_WaitingTimeAfterPotWon);
                        }
                    }
                }
            }
            GameEnded(this, new EventArgs());
            m_State = TypeState.WaitForPlayers;
            TryToBegin();
        }
        private void DecideWinners()
        {
            m_Table.CleanPotsForWinning();
            NextState(TypeState.DistributeMoney);
        }
        private void WaitALittle(int waitingTime)
        {
            Thread.Sleep(waitingTime);
        }
        private void TryToBegin()
        {
            m_Table.DecidePlayingPlayers();
            if (m_Table.NbPlaying > 1)
            {
                m_Table.InitTable();
                m_Dealer.FreshDeck();
                NextState(TypeState.WaitForBlinds);
                GameBlindNeeded(this, new EventArgs());
            }
            else if (m_Table.Players.Count > 1)
            {
                m_State = TypeState.End;
                EverythingEnded(this, new EventArgs());
            }
            else
            {
                m_Table.NoSeatDealer = -1;
                m_Table.NoSeatSmallBlind = -1;
                m_Table.NoSeatSmallBlind = -1;
            }
        }
    }
}
